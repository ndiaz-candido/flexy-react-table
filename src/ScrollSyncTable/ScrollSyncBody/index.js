import FlexyTableBody from './FlexyTableBody';

export default FlexyTableBody;
