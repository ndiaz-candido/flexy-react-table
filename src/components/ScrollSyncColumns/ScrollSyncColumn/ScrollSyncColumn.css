.scrollSyncColumn {
  padding: 10px;
}
