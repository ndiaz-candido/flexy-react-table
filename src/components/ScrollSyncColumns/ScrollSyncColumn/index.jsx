import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { scrollSyncColumn } from './index.module.css';

const ScrollSyncColumn = ({ width, children, className }) => {
  const styles = {
    width,
    flex: `0 0 ${width}`,
  };
  const classNames = classnames(className, scrollSyncColumn);

  return (
    <div className={classNames} style={styles}>
      {children}
    </div>
  );
};

ScrollSyncColumn.propTypes = {
  width: PropTypes.string,
  children: PropTypes.any.isRequired,

  /** The class name to override the default styles */
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

ScrollSyncColumn.defaultProps = {
  width: '150px',
  className: ''
};

export default ScrollSyncColumn;
