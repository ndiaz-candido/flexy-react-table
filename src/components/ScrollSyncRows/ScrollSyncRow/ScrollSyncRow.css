.scrollSyncRow {
  display: flex;
  overflow-x: hidden;
  min-height: 50px;
}

.sticky {
  position: sticky;
  top: 0;
  z-index: 1000;
}
