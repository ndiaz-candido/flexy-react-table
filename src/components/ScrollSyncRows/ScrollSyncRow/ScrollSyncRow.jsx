import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  LeftStickySection,
  RightStickySection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width,
  flex: `0 0 ${width}`,
});

class ScrollSyncRow extends PureComponent {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = Number(props.rowId);
  }

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({
    columns,
    children = [],
    columnClassName,
    isHeader,
  }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const {
        dataKey = '',
        width = ScrollSyncRow.DEFAULT_COLUMN_WIDTH,
        stickyAlign,
      } = column.props;
      const cellContent = isHeader
        ? column.props.children
        : React.Children.toArray(children).find(
            cell => cell.props.column === dataKey
          );
      const styles = cellDefaultStyles(width);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { isSticky, scrollLeft, onScroll, className } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftStickySection cells={leftStickySection} />
        <ScrollableSection
          rowId={this.rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightStickySection cells={rightStickySection} />
      </div>
    );
  }
}

ScrollSyncRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
};

ScrollSyncRow.defaultProps = {
  className: '',
};

export default ScrollSyncRow;
