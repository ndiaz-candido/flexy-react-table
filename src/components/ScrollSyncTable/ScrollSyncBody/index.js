import ScrollSyncBody from './ScrollSyncBody';

export default ScrollSyncBody;
