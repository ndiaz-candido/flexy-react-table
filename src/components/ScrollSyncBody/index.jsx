import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import ScrollSyncRow from '../ScrollSyncRows/ScrollSyncRow';

class ScrollSyncBody extends PureComponent {
  state = {
    scrollLeft: 0,
    rowBeingScrolled: null,
  };

  handleScrollEvent = (
    { isScrollingOnSameRow, rowBeingScrolled, scrollLeft },
    scrollEventCallback
  ) => {
    scrollEventCallback();

    if (isScrollingOnSameRow) {
      this.setState({ scrollLeft });
    } else {
      this.setState({ rowBeingScrolled, scrollLeft });
    }
  };

  renderHeaderRow = () => {
    const { columns, stickHeader } = this.props;
    const { rowBeingScrolled, scrollLeft } = this.state;

    return (
      <ScrollSyncRow
        key="header"
        rowId={ScrollSyncRow.HEADER_ROW_ID}
        isSticky={stickHeader}
        onScroll={this.handleScrollEvent}
        columns={columns}
        rowBeingScrolled={rowBeingScrolled}
        scrollLeft={scrollLeft}
        isHeader
      />
    );
  };

  injectPropsToRow = (row, props) => (
    <Fragment key={props.key}>{React.cloneElement(row, props)}</Fragment>
  );

  render() {
    const { rows, columns } = this.props;
    const { rowBeingScrolled, scrollLeft } = this.state;

    return (
      <Fragment>
        <Fragment key="0">{this.renderHeaderRow()}</Fragment>
        {rows.map((row, rowIndex) =>
          this.injectPropsToRow(row, {
            key: row.props.name || rowIndex,
            rowId: rowIndex + 1,
            onScroll: this.handleScrollEvent,
            columns,
            rowBeingScrolled,
            scrollLeft,
            ...row.props,
          })
        )}
      </Fragment>
    );
  }
}

ScrollSyncBody.defaultProps = {
  rows: [],
  columns: [],
};

ScrollSyncBody.propTypes = {
  /** An array of ScrollSyncRow */
  rows: PropTypes.array.isRequired,

  /** An array of ScrollSyncColumn */
  columns: PropTypes.array.isRequired,
};

export default ScrollSyncBody;
