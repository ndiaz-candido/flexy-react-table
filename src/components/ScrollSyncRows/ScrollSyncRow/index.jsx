import React, { PureComponent } from 'react';
import ScrollSyncTableCell from '../../ScrollSyncCell';
import LeftStickySection from './Sections/LeftStickySection';
import RightStickySection from './Sections/RightStickySection';
import ScrollableSection from './Sections/ScrollableSection';
import './ScrollSyncRow.css';

class ScrollSyncRow extends PureComponent {
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  rowColumns = [];
  leftStickySection = [];
  rightStickySection = [];
  scrollableSection = [];
  rowId = -1;

  constructor(props) {
    super(props);

    this.rowId = props.rowId;
  }

  componentWillMount() {
    if (this.rowId === ScrollSyncRow.HEADER_ROW_ID) {
      this.rowColumns = this.props.columns;
      this.distribuiteColumnsPerSection();
    } else {
      if (this.areChildrenSyncCells()) {
        this.rowColumns = this.setChildPerColumn();
        this.distribuiteColumnsPerSection();
      } else {
        console.warn('ScrollSyncRow: all the row children must be instances of ScrollSyncCell');
      }
    }
  }

  areChildrenSyncCells = () => {
    const { children } = this.props;
    const childrenArray = React.Children.toArray(children);

    return childrenArray.some((child) => (child.type === ScrollSyncTableCell));
  };

  replaceColumnChildren = (column, index) => {
    const { children } = this.props;
    const columnName = column.props.name;
    const scrollSyncCells = React.Children.toArray(children);
    const scrollSyncCellMatchingColumn = scrollSyncCells.find((cell) => (cell.props.column === columnName));
    const scrollSyncCellMatchingColumnChildren = scrollSyncCellMatchingColumn ? scrollSyncCellMatchingColumn.props.children : '';

    return React.cloneElement(column, { children: scrollSyncCellMatchingColumnChildren });
  }

  setChildPerColumn() {
    const { children, columns } = this.props;
    let rowColumns = [];

    if (children && React.Children.count(children) > 0) {
      rowColumns = columns.map(this.replaceColumnChildren);
    } else {
      console.warn('ScrollSyncRow: no ScrollSyncCells were specified as row children');
    }

    return rowColumns;
  }

  distribuiteColumnsPerSection() {
    this.rowColumns.forEach((column) => {
      switch (column.props.stickyAlign) {
        case 'left': this.leftStickySection.push(column); break;
        case 'rigth': this.rightStickySection.push(column); break;
        default: this.scrollableSection.push(column); break;
      }
    });
  }

  render() {
    const {
      rowId,
      isSticky,
      rowBeingScrolled,
      scrollLeft,
      onScroll
    } = this.props;

    const rowClasses = isSticky ? 'scrollSyncRow sticky' : 'scrollSyncRow';

    return (
      <div className={rowClasses}>
        <LeftStickySection columns={this.leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          columns={this.scrollableSection}
          rowBeingScrolled={rowBeingScrolled}
          scrollLeft={scrollLeft}
          onScroll={onScroll}
        />
        <RightStickySection columns={this.rightStickySection} />
      </div>
    );
  };
};

export default ScrollSyncRow;
