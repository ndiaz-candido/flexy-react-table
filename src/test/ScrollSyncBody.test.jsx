import React from 'react';
import { shallow } from 'enzyme';
import ScrollSyncBody from '../components/ScrollSyncBody';
import ScrollSyncColumn from '../components/ScrollSyncColumns/ScrollSyncColumn';
import ScrollSyncRow from '../components/ScrollSyncRows/ScrollSyncRow';

describe('ScrollSyncBody', () => {
  const columns = [
    <ScrollSyncColumn key="a" name="a" />,
    <ScrollSyncColumn key="b" name="b" />,
  ];
  const rows = [
    <ScrollSyncRow key="a" column="a" />,
    <ScrollSyncRow key="b" column="b" />,
  ];

  test('should render its default content', () => {
    const ScrollSyncBodyComponent = shallow(<ScrollSyncBody />);

    expect(ScrollSyncBodyComponent).toMatchSnapshot();
  });

  test('should render the given columns and rows', () => {
    const ScrollSyncBodyComponent = shallow(
      <ScrollSyncBody columns={columns} rows={rows} />
    );

    expect(ScrollSyncBodyComponent).toMatchSnapshot();
  });

  test('should pass the provided class names to the rows', () => {
    const ScrollSyncBodyComponent = shallow(
      <ScrollSyncBody
        columns={columns}
        rows={rows}
        rowClassName="row-class-name"
        columnClassName="column-class-name"
        headerClassName="header-class-name"
        headerColumnClassName="header-column-class-name"
      />
    );

    expect(ScrollSyncBodyComponent).toMatchSnapshot();
  });

  test('should set isSticky when stickHeader is provided', () => {
    const ScrollSyncBodyComponent = shallow(
      <ScrollSyncBody columns={columns} rows={rows} stickHeader />
    );

    expect(ScrollSyncBodyComponent).toMatchSnapshot();
  });

  test('should update the scroll left position when handleScrollEvent is called', () => {
    const ScrollSyncBodyComponent = shallow(
      <ScrollSyncBody columns={columns} rows={rows} />
    );
    const HeaderRow = ScrollSyncBodyComponent.find('ScrollSyncRow').first();

    HeaderRow.prop('onScroll')({ scrollLeft: 10 });

    expect(ScrollSyncBodyComponent.state()).toEqual({
      scrollLeft: 10,
    });
  });

  test('should update the row being scrolled when handleScrollEvent is called from a different row', () => {
    const ScrollSyncBodyComponent = shallow(
      <ScrollSyncBody columns={columns} rows={rows} />
    );

    ScrollSyncBodyComponent.setState({
      scrollLeft: 10,
    });

    const HeaderRow = ScrollSyncBodyComponent.find('ScrollSyncRow').first();

    HeaderRow.prop('onScroll')({ scrollLeft: 20 });

    expect(ScrollSyncBodyComponent.state()).toEqual({
      scrollLeft: 20,
    });
  });
});
