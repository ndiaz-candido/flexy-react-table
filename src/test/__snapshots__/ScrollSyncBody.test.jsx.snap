// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollSyncBody should pass the provided class names to the rows 1`] = `
<React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className="header-class-name"
      columnClassName="header-column-class-name"
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      isHeader={true}
      isSticky={false}
      key="header"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={0}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className="row-class-name"
      column="a"
      columnClassName="column-class-name"
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="0"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={1}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="1"
  >
    <ScrollSyncRow
      className="row-class-name"
      column="b"
      columnClassName="column-class-name"
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="1"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={2}
      scrollLeft={0}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`ScrollSyncBody should render its default content 1`] = `
<React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className=""
      columnClassName=""
      columns={Array []}
      isHeader={true}
      isSticky={false}
      key="header"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={0}
      scrollLeft={0}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`ScrollSyncBody should render the given columns and rows 1`] = `
<React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className=""
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      isHeader={true}
      isSticky={false}
      key="header"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={0}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className=""
      column="a"
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="0"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={1}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="1"
  >
    <ScrollSyncRow
      className=""
      column="b"
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="1"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={2}
      scrollLeft={0}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`ScrollSyncBody should set isSticky when stickHeader is provided 1`] = `
<React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className=""
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      isHeader={true}
      isSticky={true}
      key="header"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={0}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="0"
  >
    <ScrollSyncRow
      className=""
      column="a"
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="0"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={1}
      scrollLeft={0}
    />
  </React.Fragment>
  <React.Fragment
    key="1"
  >
    <ScrollSyncRow
      className=""
      column="b"
      columnClassName=""
      columns={
        Array [
          <ScrollSyncColumn
            className=""
            name="a"
            width="150px"
          />,
          <ScrollSyncColumn
            className=""
            name="b"
            width="150px"
          />,
        ]
      }
      key="1"
      onScroll={[Function]}
      rowBeingScrolled={null}
      rowId={2}
      scrollLeft={0}
    />
  </React.Fragment>
</React.Fragment>
`;
